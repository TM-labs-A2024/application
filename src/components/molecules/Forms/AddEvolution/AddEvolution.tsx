import { ArrowBackIcon } from '@chakra-ui/icons'
import {
  Button,
  FormControl,
  FormErrorMessage,
  Input,
  Stack,
  IconButton,
  Text,
  Textarea
} from '@chakra-ui/react'
import { evolutionTypes } from '@src/constants'
import { ReactSelectOption } from '@src/types'
import { isIOS } from '@utils/index'
import { useRouter } from 'next/navigation'
import React, { ReactElement } from 'react'
import { Controller, useForm, FieldErrors } from 'react-hook-form'
import Select from 'react-select'

const userTypeOptions = evolutionTypes.map((option) => ({
  value: option.id,
  label: option.name
}))

type FormData = {
  type: ReactSelectOption | null
  reason: string
  description: string
  history: string
  examination: string
  comments: string
}

export default function AddEvolutionForm(): ReactElement {
  // --- Hooks -----------------------------------------------------------------
  const {
    handleSubmit,
    formState: { errors, isSubmitting },
    register,
    control
  } = useForm<FormData>()

  const router = useRouter()
  // --- END: Hooks ------------------------------------------------------------

  // --- Local state -----------------------------------------------------------
  // --- END: Local state ------------------------------------------------------

  // --- Refs ------------------------------------------------------------------
  // --- END: Refs -------------------------------------------------------------

  // --- Redux -----------------------------------------------------------------
  // --- END: Redux ------------------------------------------------------------

  // --- Side effects ----------------------------------------------------------
  // --- END: Side effects -----------------------------------------------------

  // --- Data and handlers -----------------------------------------------------
  const onSubmitDetails = (data: FormData) => {
    alert(JSON.stringify(data))
  }

  const verifyErrors = (errors: FieldErrors<FormData>) => Object.keys(errors).length > 0
  // --- END: Data and handlers ------------------------------------------------

  return (
    <div
      className={`flex h-screen w-screen flex-col overflow-hidden p-8 lg:px-96 ${isIOS() ? 'pt-20' : 'pt-8'}`}
      data-testid="add-evolution-form"
    >
      <div className="mb-8 flex flex-row items-center justify-start gap-4">
        <IconButton
          size="xl"
          aria-label="back"
          variant="link"
          icon={<ArrowBackIcon />}
          onClick={() => {
            router.back()
          }}
        />
        <Text className="font-medium">Nueva evolución</Text>
      </div>
      <form onSubmit={handleSubmit(onSubmitDetails)} className="h-full w-full">
        <FormControl
          isInvalid={verifyErrors(errors)}
          className="relative flex h-full w-full flex-col justify-between"
        >
          <Stack spacing={4} className="h-5/6 overflow-scroll">
            <Controller
              control={control}
              name="type"
              rules={{
                required: 'Este campo es obligatorio'
              }}
              render={({ field }) => (
                <Select id="type" {...field} placeholder="Tipo" options={userTypeOptions} />
              )}
            />
            <FormErrorMessage>{errors?.type && errors?.type?.message}</FormErrorMessage>
            <Input
              id="reason"
              placeholder="Patología"
              className="min-h-10"
              {...register('reason', {
                required: 'Este campo es obligatorio'
              })}
            />
            <FormErrorMessage>{errors?.reason && errors?.reason?.message}</FormErrorMessage>
            <label className="-mb-2" htmlFor="description">
              Descripción (Notas evolutivas)
            </label>
            <Textarea
              id="description"
              {...register('description', {
                required: 'Este campo es obligatorio'
              })}
            />
            <FormErrorMessage>
              {errors?.description && errors?.description?.message}
            </FormErrorMessage>
            <label className="-mb-2" htmlFor="history">
              Antecedentes
            </label>
            <Textarea
              id="history"
              {...register('history', {
                required: 'Este campo es obligatorio'
              })}
            />
            <FormErrorMessage>{errors?.history && errors?.history?.message}</FormErrorMessage>
            <label className="-mb-2" htmlFor="examination">
              Exámenes físicos
            </label>
            <Textarea
              id="examination"
              {...register('examination', {
                required: 'Este campo es obligatorio'
              })}
            />
            <FormErrorMessage>
              {errors?.examination && errors?.examination?.message}
            </FormErrorMessage>
            <label className="-mb-2" htmlFor="comments">
              Comentarios
            </label>
            <Textarea id="comments" {...register('comments')} />
          </Stack>
          <Button isLoading={isSubmitting} type="submit" className="absolute bottom-12">
            Guardar
          </Button>
        </FormControl>
      </form>
    </div>
  )
}
