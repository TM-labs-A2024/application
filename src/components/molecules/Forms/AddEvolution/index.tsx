import AddEvolutionForm from './AddEvolution'

export default AddEvolutionForm
