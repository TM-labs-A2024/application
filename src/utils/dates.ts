export const formatDate = (date: string) => {
  const chunks = date.split('-')

  const newDate = chunks[1] + '-' + chunks[2] + '-' + chunks[0]

  return newDate
}
