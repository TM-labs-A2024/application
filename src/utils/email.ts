import emailjs from '@emailjs/browser'

export function sendEmail(data: {
  from_name: string
  to_name: string
  code: string | number
  to_email: string
}) {
  const templateParams = {
    ...data
  }

  const serviceKey = process.env.NEXT_PUBLIC_EMAIL_SERVICE_KEY ?? ''
  const templateKey = process.env.NEXT_PUBLIC_EMAIL_TEMPLATE_KEY ?? ''
  const apiKey = process.env.NEXT_PUBLIC_EMAIL_API_KEY ?? ''

  emailjs.send(serviceKey, templateKey, templateParams, apiKey).catch(() => {
    alert('Algo sali√≥ mal, por favor reintenta el registro.')
  })
}
