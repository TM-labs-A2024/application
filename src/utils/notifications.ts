import { GENERIC_ERROR } from '@constants/index'
import { isMobile } from '@utils/index'
import { Store } from 'react-notifications-component'

export const setupErrorNotification = () => {
  Store.addNotification(
    GENERIC_ERROR('Algo salió mal, por favor inténtelo de nuevo.', isMobile(window))
  )
}
