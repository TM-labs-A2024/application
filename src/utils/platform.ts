import { Capacitor } from '@capacitor/core'

export const isWeb = () => Capacitor.getPlatform() === 'web'

export const isIOS = () => Capacitor.getPlatform() === 'ios'

export const isAndroid = () => Capacitor.getPlatform() === 'android'
