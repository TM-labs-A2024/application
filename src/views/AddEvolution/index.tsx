import AddEvolution from '@components/molecules/Forms/AddEvolution'
import { EvolutionFormData } from '@src/types'
import React from 'react'

export default function AddEvolutionView({
  context
}: {
  context: { onSubmit: (data: EvolutionFormData) => void; isLoading: boolean }
}) {
  return <AddEvolution context={context} />
}
