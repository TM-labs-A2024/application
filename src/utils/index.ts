import { formatDate } from './dates'
import { sendEmail, generateVerificationCode, setupEmailSending } from './email'
import { setupErrorNotification } from './notifications'
import { isWeb, isIOS, isAndroid, isMobile } from './platform'

export {
  isWeb,
  isIOS,
  isAndroid,
  isMobile,
  sendEmail,
  generateVerificationCode,
  setupEmailSending,
  setupErrorNotification,
  formatDate
}
