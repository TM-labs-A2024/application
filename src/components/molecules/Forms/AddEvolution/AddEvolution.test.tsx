import { render, screen } from '@test/utils'
import 'next/navigation'

import AddEvolution from './AddEvolution'
import '@testing-library/jest-dom'

jest.mock('next/navigation')

describe('Molecules > Forms > AddEvolution test', () => {
  test('The component requires the fields: Type, reason, description, history, examination and comments', () => {
    render(<AddEvolution />)

    expect(screen.getByText('Tipo')).toBeInTheDocument()
    expect(screen.getByPlaceholderText('Patología')).toBeInTheDocument()
    expect(screen.getByLabelText('Descripción (Notas evolutivas)')).toBeInTheDocument()
    expect(screen.getByLabelText('Antecedentes')).toBeInTheDocument()
    expect(screen.getByLabelText('Exámenes físicos')).toBeInTheDocument()
    expect(screen.getByLabelText('Comentarios')).toBeInTheDocument()
  })

  test('Matches the snapshot', () => {
    render(<AddEvolution />)

    expect(screen.getByTestId('add-evolution-form')).toMatchSnapshot()
  })
})
