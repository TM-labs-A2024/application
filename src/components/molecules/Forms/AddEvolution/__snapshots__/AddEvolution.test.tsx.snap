// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Molecules > Forms > AddEvolution test Matches the snapshot 1`] = `
<div
  class="flex h-screen w-screen flex-col overflow-hidden p-8 lg:px-96 pt-8"
  data-testid="add-evolution-form"
>
  <div
    class="mb-8 flex flex-row items-center justify-start gap-4"
  >
    <button
      aria-label="back"
      class="chakra-button css-84vx7q"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="chakra-icon css-onkibi"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M20 11H7.83l5.59-5.59L12 4l-8 8 8 8 1.41-1.41L7.83 13H20v-2z"
          fill="currentColor"
        />
      </svg>
    </button>
    <p
      class="chakra-text font-medium css-0"
    >
      Nueva evolución
    </p>
  </div>
  <form
    class="h-full w-full"
  >
    <div
      class="chakra-form-control relative flex h-full w-full flex-col justify-between css-1kxonj9"
      role="group"
    >
      <div
        class="chakra-stack h-5/6 overflow-scroll css-1811skr"
      >
        <div
          class=" css-b62m3t-container"
          id="type"
        >
          <span
            class="css-1f43avz-a11yText-A11yText"
            id="react-select-3-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            class="css-1f43avz-a11yText-A11yText"
            role="log"
          />
          <div
            class=" css-13cymwt-control"
          >
            <div
              class=" css-1fdsijx-ValueContainer"
            >
              <div
                class=" css-1jqq78o-placeholder"
                id="react-select-3-placeholder"
              >
                Tipo
              </div>
              <div
                class=" css-qbdosj-Input"
                data-value=""
              >
                <input
                  aria-activedescendant=""
                  aria-autocomplete="list"
                  aria-describedby="react-select-3-placeholder"
                  aria-expanded="false"
                  aria-haspopup="true"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  class=""
                  id="react-select-3-input"
                  role="combobox"
                  spellcheck="false"
                  style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class=" css-1hb7zxy-IndicatorsContainer"
            >
              <span
                class=" css-1u9des2-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                class=" css-1xc3v61-indicatorContainer"
              >
                <svg
                  aria-hidden="true"
                  class="css-tj5bde-Svg"
                  focusable="false"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
          <input
            name="type"
            type="hidden"
            value=""
          />
        </div>
        <input
          class="chakra-input min-h-10 css-1cjy4zv"
          id="reason"
          name="reason"
          placeholder="Patología"
        />
        <label
          class="-mb-2"
          for="description"
        >
          Descripción (Notas evolutivas)
        </label>
        <textarea
          class="chakra-textarea css-1msuebi"
          id="description"
          name="description"
        />
        <label
          class="-mb-2"
          for="history"
        >
          Antecedentes
        </label>
        <textarea
          class="chakra-textarea css-1msuebi"
          id="history"
          name="history"
        />
        <label
          class="-mb-2"
          for="examination"
        >
          Exámenes físicos
        </label>
        <textarea
          class="chakra-textarea css-1msuebi"
          id="examination"
          name="examination"
        />
        <label
          class="-mb-2"
          for="comments"
        >
          Comentarios
        </label>
        <textarea
          class="chakra-textarea css-1msuebi"
          id="comments"
          name="comments"
        />
      </div>
      <button
        class="chakra-button absolute bottom-12 css-ez23ye"
        type="submit"
      >
        Guardar
      </button>
    </div>
  </form>
</div>
`
